### ISSUE TITLE
More verbose error messages in `oneOf`

### ISSUE THREAD

**Issue by kubukoz** [SystemFw/dynosaur#9]
I like the simplicity of the current error model (just a string is sometimes exactly what you want), but currently if a union fails to decode, you get virtually no information.

Combining the errors somehow would be great to have. Some ideas:

- Either -> EitherNel (probably the most intrusive change)
- ReadError as an ADT
- just combining the strings to a larger string in the decoding logic for alternatives

**Comment by SystemFw** [comment]
I actually have some notes on errors in general, the design space is not so trivial. I'm gonna share when I get some time

**Comment by kubukoz** [comment]
Agreed, it's definitely not trivial 😅 especially if we consider error accumulation for records too.

**Comment by kubukoz** [comment]
yeah, just hit this again - @SystemFw do you have bandwidth to give errors a closer look?

For @semenodm's toString issue we can probably to a quick workaround by adding a `toString` override.

**Comment by SystemFw** [comment]
I'd take a PR to fix the immediate problem with `toString`, I don't know how long it'll be before I can look at the design more broadly

**Comment by filosganga** [comment]
I believe the #29 has addressed the issue partially: the resulting error now contains all the underlying branches failures
